controllers/fenceagentsremediation_controller_test.go

```diff
@@ -0,0 +1,195 @@
+				Eventually(func() (bool, error) {
+					res, err := cliCommandsEquality(underTestFAR)
+					return res, err
+				}, 1*time.Second, 500*time.Millisecond).Should(BeTrue(), BeEmpty())
+			})
+		})
+	})
+})
+
+// buildFARParameters from string to arrays to two string maps (key-value manner)
+func buildFARParameters(fields []string, values []string, nodeFields []string, nodeValues []string, node string) (map[v1alpha1.ParameterName]string, map[v1alpha1.ParameterName]map[v1alpha1.NodeName]string) {
+	testShareParam := make(map[v1alpha1.ParameterName]string)
+	testNodeParam := make(map[v1alpha1.ParameterName]map[v1alpha1.NodeName]string)
+	i := 0
+	for i = 0; i < len(values); i++ {
+		field := v1alpha1.ParameterName(fields[i])
+		testShareParam[field] = values[i]
+	}
+
+	nodeName := v1alpha1.NodeName(node)
+	numNodeParam := len(fields) - len(values)
+	for j := 0; j < numNodeParam; j++ {
+		if indexOf(node, nodeFields) > -1 {
+			field := v1alpha1.ParameterName(fields[i+j])
+			testNodeParam[field] = make(map[v1alpha1.NodeName]string)
+			testNodeParam[field][nodeName] = nodeValues[indexOf(node, nodeFields)]
+		}
+	}
+	return testShareParam, testNodeParam
+}
+
+// indexOf return the index of element in data array. If it is not found, return -1
+func indexOf(element string, data []string) int {
+	for i, v := range data {
+		if element == v {
+			return i
+		}
+	}
+	return -1
+}
+
+// newFenceAgentsRemediation assign the input to the FenceAgentsRemediation's Spec
+func newFenceAgentsRemediation(nodeName string, agent string, sharedparameters map[v1alpha1.ParameterName]string, nodeparameters map[v1alpha1.ParameterName]map[v1alpha1.NodeName]string) *v1alpha1.FenceAgentsRemediation {
+	return &v1alpha1.FenceAgentsRemediation{
+		ObjectMeta: metav1.ObjectMeta{Name: nodeName, Namespace: defaultNamespace},
+		Spec: v1alpha1.FenceAgentsRemediationSpec{
+			Agent:            agent,
+			SharedParameters: sharedparameters,
+			NodeParameters:   nodeparameters,
+		},
+	}
+}
+
+// buildFarPod builds a dummy pod with FAR label and namespace
+func buildFarPod() *corev1.Pod {
+	fenceAgentsPod := &corev1.Pod{}
+	fenceAgentsPod.Labels = faPodLabels
+	fenceAgentsPod.Name = "mock-fence-agents"
+	fenceAgentsPod.Namespace = defaultNamespace
+	container := corev1.Container{
+		Name:  "foo",
+		Image: "foo",
+	}
+	fenceAgentsPod.Spec.Containers = []corev1.Container{container}
+	return fenceAgentsPod
+}
+
+// cliCommandsEquality creates the command for CLI and compares it with the production command
+func cliCommandsEquality(far *v1alpha1.FenceAgentsRemediation) (bool, error) {
+	//fence_ipmilan --ip=192.168.111.1 --ipport=6233 --username=admin --password=password --action=status --lanplus
+	if mocksExecuter.command == nil {
+		return false, errors.New("executedCommand is null")
+	}
+	expectedCommand, err := buildFenceAgentParams(far)
+	if err != nil {
+		return false, err
+	}
+	expectedCommand = append([]string{far.Spec.Agent}, expectedCommand...)
```

Comment:
We use the same code for building the command here in the test and in the controller. So IIUC the only thing which is tested here is whether Execute was called or not. It's impossible that the command is not the same.

I would like to see a that buildFenceAgentParams() is tested as well. One possibily is to hardcode the expected command here instead of calling buildFenceAgentParams()

Reply:
> I would like to see that buildFenceAgentParams() is tested as well. One possibily is to hardcode the expected command here instead of calling buildFenceAgentParams()

I am not sure I follow. In the beginning I understand that you are suggesting to call buildFenceAgentParams() and then you suggest to hardcode the expected command and don't use that function.

Reply:
> expectedCommand, err := buildFenceAgentParams(far)
> 	if err != nil {
> 		return false, err
> 	}

If the intention was to still using this function and then hardcode the expected command, so I would keep the above lines to test _buildFenceAgentParams()_  and then explicitly assign the expected command for the _expectedCommand_ variable as below:

 ```suggestion
	// hardcode expected command
	expectedCommand = []string{"","--lanplus","--password=password","--username=admin","--action=reboot","--ip=192.168.111.1","--ipport=6233"}
```

Did I get it right?

Reply:
> you are suggesting to call buildFenceAgentParams()
sorry if I did so (can't remember), I probably didn't understand the complete test yet because of all the confusion around FAR and FARTemplate.
The change now looks almost good, I think you need to add the agent's name as first string value (also hardcoded)?